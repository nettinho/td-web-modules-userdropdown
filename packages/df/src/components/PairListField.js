import React from "react";
import PropTypes from "prop-types";
import { Form, Button, Icon } from "semantic-ui-react";
import { injectIntl } from "react-intl";

export const PairListField = ({
  intl: { formatMessage },
  label,
  name,
  required,
  value,
  onChange
}) => {
  const handleAddUrl = e => {
    e && e.preventDefault();
    onChange(e, { name, value: [...value, { url_name: "", url_value: "" }] });
  };

  const handleRemoveUrl = (e, { id }) => {
    e && e.preventDefault();

    const newValue = [...value];
    newValue.splice(id, 1);
    onChange(e, { name, value: newValue });
  };

  const doOnChange = (e, data) => {
    const newValue = [...value];
    const idx = data.id;
    newValue[idx] = { ...newValue[idx], [data.name]: data.value };
    onChange(e, { name, value: newValue });
  };

  return (
    <Form.Field required={required}>
      <label>{label}</label>
      {value.map((v, i) => (
        <Form.Group widths="equal" key={i}>
          <Form.Input
            id={i}
            required={required}
            value={v.url_name}
            name="url_name"
            onChange={doOnChange}
            placeholder={formatMessage({
              id: "fields.pairlist.url.name.placeholder"
            })}
          />
          <Form.Input
            id={i}
            required={required}
            value={v.url_value}
            name="url_value"
            onChange={doOnChange}
            placeholder={formatMessage({
              id: "fields.pairlist.url.value.placeholder"
            })}
          />
          <Button
            name="delete_buttom"
            compact
            className="no_background"
            onClick={e => handleRemoveUrl(e, i)}
          >
            <Icon name="minus" />
          </Button>
        </Form.Group>
      ))}
      <Button
        name="add_buttom"
        compact
        className="no_background"
        onClick={handleAddUrl}
      >
        <Icon name="add" color="green" />
      </Button>
    </Form.Field>
  );
};

PairListField.propTypes = {
  label: PropTypes.string,
  name: PropTypes.string,
  required: PropTypes.bool,
  value: PropTypes.array,
  onChange: PropTypes.func.isRequired,
  handleAddUrl: PropTypes.func,
  handleRemoveUrl: PropTypes.func
};

export default injectIntl(PairListField);
