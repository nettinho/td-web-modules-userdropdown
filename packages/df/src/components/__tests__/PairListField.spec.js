import React from "react";
import { shallowWithIntl } from "intl-stub";
import { PairListField } from "../PairListField";

describe("<PairListField />", () => {
  const label = "label";
  const name = "name";
  const required = true;
  const value = [{ url_name: "Foo", url_value: "Bar" }];
  const onChange = jest.fn();
  const props = { label, name, required, value, onChange };

  it("matches the latest snapshot", () => {
    const wrapper = shallowWithIntl(<PairListField {...props} />);
    expect(wrapper).toMatchSnapshot();
  });

  it("generates correct onChange value format when changing", () => {
    const data = {
      id: 0,
      name: "url_name",
      value: "data_value"
    };
    const component = shallowWithIntl(<PairListField {...props} />);
    component.find({ name: "url_name" }).simulate("change", null, data);
    expect(onChange).toBeCalledWith(null, {
      name: "name",
      value: [{ url_name: "data_value", url_value: "Bar" }]
    });
  });

  it("generates row when add buttom is pressed", () => {
    const event = { preventDefault() {} };
    const component = shallowWithIntl(<PairListField {...props} />);
    component.find({ name: "add_buttom" }).simulate("click", event);
    expect(onChange).toBeCalledWith(event, {
      name: "name",
      value: [
        { url_name: "Foo", url_value: "Bar" },
        { url_name: "", url_value: "" }
      ]
    });
  });

  it("deletes row when delete buttom is pressed", () => {
    const event = { preventDefault() {} };
    const component = shallowWithIntl(<PairListField {...props} />);
    component.find({ name: "delete_buttom" }).simulate("click", event);
    expect(onChange).toBeCalledWith(event, { name: "name", value: [] });
  });
});
