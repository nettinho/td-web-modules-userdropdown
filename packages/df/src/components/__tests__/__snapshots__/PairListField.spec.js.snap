// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PairListField /> matches the latest snapshot 1`] = `
<FormField
  required={true}
>
  <label>
    label
  </label>
  <FormGroup
    key="0"
    widths="equal"
  >
    <FormInput
      as={[Function]}
      control={[Function]}
      id={0}
      name="url_name"
      onChange={[Function]}
      placeholder="fields.pairlist.url.name.placeholder"
      required={true}
      value="Foo"
    />
    <FormInput
      as={[Function]}
      control={[Function]}
      id={0}
      name="url_value"
      onChange={[Function]}
      placeholder="fields.pairlist.url.value.placeholder"
      required={true}
      value="Bar"
    />
    <Button
      as="button"
      className="no_background"
      compact={true}
      name="delete_buttom"
      onClick={[Function]}
      role="button"
    >
      <Icon
        as="i"
        name="minus"
      />
    </Button>
  </FormGroup>
  <Button
    as="button"
    className="no_background"
    compact={true}
    name="add_buttom"
    onClick={[Function]}
    role="button"
  >
    <Icon
      as="i"
      color="green"
      name="add"
    />
  </Button>
</FormField>
`;
